Change cheatsheet to clarify some alpaka structures (#2241)
alpaka-group/alpaka: docs/source/basic/cheatsheet.rst

Allocate a buffer in host memory
  .. code-block:: c++

>      // Use alpaka vector as a static array for the extent(dimensions)

Reviewer: ```suggestion
     // Use alpaka vector as a static array for the extent(dimensions)
```
I prefer the word extents over dimensions. In my mental model, if you have a 2D array of size 10x12 elements, it has 2 dimensions and extents 10 and 12. So dimensions for me, are the number of extents.

Author: No. When we say "dimensions of an A4 paper" or "dimensions of a box" we don't mean "2 dimensional" or "3 dimentional" we just mean height,width and depth values.

Reviewer 2: Well, an A4 paper has 2 dimensions, and those dimensions are 29.7 cm and 21 cm.

To be pedantic, one could say that an A4 paper has a dimensionality of 2, and its size along those dimensions are 29.7 cm and 21 cm.

In any case, I don't know if there is an official nomenclature in C/C++.
The STL has traits for them, and uses [rank](https://en.cppreference.com/w/cpp/types/rank) for the number of dimensions, and [extent](https://en.cppreference.com/w/cpp/types/extent) the size along each dimension.

Author: _Extent_ and _pitch_ are Cuda concepts mainly, C++ developers know "dimensions", "padding", "alignment".  Therefore I have just wanted to add a few words to clarify.

Author: Ok I will do "array of dimensions" :)

Reviewer: In addition to what @fwyzard said, there is also [`std::extents`](https://en.cppreference.com/w/cpp/container/mdspan/extents).

Reviewer 2: since C++23 ... I still live in C++17 :-)

Author: 1. I think you mix up "number of dimensions" and the "dimensions". 

2. I haven't replaced or proposed replacing the term extent with the
dimensions, just tried to make the document more understandable for the
people who are new to alpaka.

Reviewer: While this mostly counts for me as well, as library designers or architects, we need to stay current with the technical evolution and design accordingly. I also want my libraries to fit the future ;)

Author: Alpaka uses of Dim and TDim, which are misleading I think. For example Dim can be NDim or NumDim ( e.g. matlab or python function names to get number of dimensions)

Author: Lets agree on this. Number of Dimensions = Number of Extents. Dimensions could be misleading as you said. I accept. Why, because it could mean a set of dimension names like {XDimension,YDimension,ZDimension} or it could mean as in daily usage {width,depth, height}.  I remove it.

Reviewer 2: > Number of Dimensions = Number of Extents.

Agreed.
It is also called "rank" (in C++) or "dimensionality".

>      // Use alpaka vector as a static array for the extent(the array of dimensions)

Reviewer: ```suggestion
     // Use alpaka vector as a static array for the extents (the array of dimensions)
```

     alpaka::Vec<Dim, Idx> extent = value;

>      // Allocate memory for the alpaka buffer which is a dynamic array

Reviewer: ```suggestion
     // Allocate memory for the alpaka buffer, which is a dynamic array
```

>      using BufHost = alpaka::Buf<DevHost, DataType, Dim, Idx>;

Reviewer: Be careful to keep the cheatsheet short! It is rather not an introduction, but really a cheatsheet for the API. At least that's my understanding.

Author: Sure, these are really important explanations. I am not planning to add more.

  .. code-block:: c++

     using Dim = alpaka::DimInt<1u>;

>      // create an alpaka vector which is a static array

Reviewer: Please use a consistent case when starting a comment
```suggestion
     // Create an alpaka vector which is a static array
```

     // Origin: Grid, Block, Thread
     // Unit: Blocks, Threads, Elems

> Access components of and destructuremulti-dimensional indices and extents (dimensions)

Reviewer: ```suggestion
Access components of and destructure multi-dimensional indices and extents (dimensions)
```
